### PR TITLE
Add event subscriptions for channel + gateway, link packet graphs to gql

### PR DESCRIPTION
**Pull request by vicmgs**

### CODE REVIEW

File: lib/console_web/controllers/event_controller.ex [hunk 1]
```diff
@@ -68,5 +68,11 @@ defmodule ConsoleWeb.EventController do
     if event.device do
       Absinthe.Subscription.publish(ConsoleWeb.Endpoint, event, event_added: event.device.id)
     end
+    if event.gateway do
+      Absinthe.Subscription.publish(ConsoleWeb.Endpoint, event, event_added: event.gateway.id)
+    end
+    if event.channel do
+      Absinthe.Subscription.publish(ConsoleWeb.Endpoint, event, event_added: event.channel.id)
```
Comment (allenan) [inline lib/console_web/controllers/event_controller.ex]: do we need some more context here? we're just using id it looks like

File: assets/js/components/events/EventsTablePaginated.jsx [hunk 2]
```diff
@@ -174,8 +174,8 @@ const query = gql`
 
 // TODO create event fragment?
```
Comment (allenan) [inline assets/js/components/events/EventsTablePaginated.jsx]: we're now using the event fields in a few places. could be worth looking into creating a fragment for it: https://www.apollographql.com/docs/react/advanced/fragments.html

File: assets/js/components/common/PacketGraph.jsx [hunk 3]
```diff
@@ -83,17 +87,13 @@ class PacketGraph extends Component {
...
-      (this.props.data[0] && this.state.data.datasets.length == 0) ||
-      (this.props.data[0] && prevProps.data[0] && this.props.data[this.props.data.length-1].id !== prevProps.data[prevProps.data.length-1].id)
-    ) {
+    if (!prevProps.data.recentEvents && recentEvents) {
       clearInterval(this.chartUpdateInterval)
-      this.updateChart(this.props.data)
+      this.updateChart(recentEvents)
       this.chartUpdateInterval = setInterval(() => {
-        this.updateChart(this.props.data)
+        this.updateChart(recentEvents)
```
Comment (allenan) [inline assets/js/components/common/PacketGraph.jsx]: is this just polling that endpoint? could we use subscriptions here so we hit the gql endpoint once, grab all recent events, and then when new ones come in, add to the graph?

Comment (vicmgs) [reply assets/js/components/common/PacketGraph.jsx]: updateChart just transforms props to state for the chart Component, it doesn't hit the endpoint.

File: lib/console_web/controllers/event_controller.ex (hunk 1 above)
Comment (vicmgs) [reply lib/console_web/controllers/event_controller.ex]: don't think so? the if statement should router the new event to the proper context. Even if the event belongs to more than 1 context, we should publish for each context as well right?

**Comment by vicmgs** [comment]
Fragmented out events gql stuff.

Also, added code for subscription in PacketGraph.jsx

File: assets/js/graphql/events.js [hunk 4]
```diff
@@ -0,0 +1,21 @@
+import gql from 'graphql-tag';
+
+export const EventFragment = gql`
```
Comment (allenan) [inline assets/js/graphql/events.js]: I usually think of camel-cased variables in javascript to mean classes. should this be a screaming snake case like `EVENTS_SUBSCRIPTION`?

File: lib/console_web/controllers/event_controller.ex (hunk 1 above)
Comment (allenan) [reply lib/console_web/controllers/event_controller.ex]: the only difference i'm seeing between these 3 publishes is the `event_added: event.channel.id` vs `event_added: event.gateway.id` which is the id of the event parent.

File: lib/console_web/schema/schema.ex [hunk 5]
```diff
@@ -124,14 +111,21 @@ defmodule ConsoleWeb.Schema do
...
   end
 
   subscription do
     field :event_added, :event do
-      arg :device_id, :string
+      arg :context_id, :string
 
       config fn args, _ ->
-        {:ok, topic: args.device_id}
+        {:ok, topic: args.context_id}
```
Comment (allenan) [inline lib/console_web/schema/schema.ex]: yeah the subscription is just taking a context_id, so maybe we should make the topic like `"#{args.context_name}/#{args.context_id}"`

**Pull request merged**